Add Redis instrumentation query sanitization

tombruijn opened the pull request:
# Description

Add a query sanitizer to the Redis instrumentation. This can be disabled
with the `sanitize_query = False` config option.

Given the query `SET key value`, the sanitized query becomes `SET ? ?`.
Both the keys and values are sanitized, as both can contain PII data.

The Redis queries are sanitized by default. This changes the default
behavior of this instrumentation. Previously it reported unsanitized
Redis queries.

This was previously discussed in the previous implementation of this PR
in PR #1571

Closes #1548

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] `tox -e test-instrumentation-redis`

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

avzis commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/__init__.py:
```diff
@@ -64,6 +64,8 @@ async def redis_get():
 response_hook (Callable) - a function with extra user-defined logic to be performed after performing the request
 this function signature is: def response_hook(span: Span, instance: redis.connection.Connection, response) -> None
 
+sanitize_query True(Default) or False - enable the Redis query sanization
```
I'd say Boolean type, and fix naming to sanitization, something like:
`sanitize_query (Boolean) - default True, enable the Redis query sanitization `

tombruijn replied:
I've updated the documentation 👍

avzis commented on instrumentation/opentelemetry-instrumentation-redis/tests/test_redis.py:
```diff
@@ -148,6 +148,45 @@ def request_hook(span, conn, args, kwargs):
+        connection = redis.connection.Connection()
+        redis_client.connection = connection
+
+        RedisInstrumentor().uninstrument()
+        RedisInstrumentor().instrument(
+            tracer_provider=self.tracer_provider
+        )
```
I think you don't need this part, it is already instrumented properly

tombruijn replied:
I've removed the unnecessary setup code here.

avzis commented:
this looks great,
will solve #1548

tombruijn commented:
I fixed the lint issue that failed the build previously. I couldn't get the linter to run locally first, but I confirmed it now passes locally.

srikanthccv commented on instrumentation/opentelemetry-instrumentation-redis/src/opentelemetry/instrumentation/redis/util.py:
```diff
@@ -48,25 +48,29 @@ def _extract_conn_attributes(conn_kwargs):
+    """Format and sanitize command arguments, and trim them as needed"""
+    if sanitize_query:
+        # Sanitized query format: "COMMAND ? ?"
+        out = [str(args[0])] + ["?"] * (len(args) - 1)
```
Will there ever be a case where args[0] raises index error?

tombruijn replied:
Not that I know off. I assume index 0 is always the Redis command that's being performed. I can add more safeguards if that helps.

srikanthccv replied:
That would be good because we don't want to instrumentation to raise exceptions ever. Users would rather lose telemetry than seeing the failures from instrumentation.

shalevr commented:
I'm not sure why to change the default behavior, 
you need to update the test here https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/tests/opentelemetry-docker-tests/tests/redis/test_redis_functional.py
and other users have to change their code too or get different results than expected

tombruijn commented:
> I'm not sure why to change the default behavior,

This was briefly discussed in the previous implementation of this, in PR https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1571.

I think sanitizing the query is a good default. It's better because users won't send any Personal Identifying Information by default, and need to explicitly enable full query reporting if they really need it.

> you need to update the test here https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/tests/opentelemetry-docker-tests/tests/redis/test_redis_functional.py
and other users have to change their code too or get different results than expected

Okay, sounds good. I'm just running into another issue where I can't run this test suite. I'm getting connection issues for something. I'll have a look.

tombruijn commented:
@shalevr I've updated the functional tests, added new tests to test `sanitize_query=False` and made sure the `db.statement` is not longer than 1000 chars. Let me know what you think.

shalevr commented on tests/opentelemetry-docker-tests/tests/redis/test_redis_functional.py:
```diff
@@ -90,6 +149,27 @@ def test_pipeline_traced(self):
+            pipeline.immediate_execute_command("SET", "b", 2)
+            pipeline.execute()
+
+        spans = self.memory_exporter.get_finished_spans()
```
I suggest to keep the old function tests and just move sanitize_query=False to them
so the test without sanitizing will not change.
and add new function to check sanitizing like you did.
BTW there is an open issue in the spesification if the default db.statement sanitization should be True/False 
https://github.com/open-telemetry/opentelemetry-specification/issues/3104

tombruijn commented:
@srikanthccv with the discussion about query sanitization by default ongoing in the [spec repo](https://github.com/open-telemetry/opentelemetry-specification/issues/3104), shall I change this PR to not sanitize the queries by default, but make the option available to sanitize query opt-in?

srikanthccv commented:
Yes, please do.

tombruijn commented:
@srikanthccv I've changed the `sanitize_query` config option to be False by default and resolved the merge conflict with the main branch. Can you retry the build? (I ran the tests, black and flake8 locally, so they should be good.)

tombruijn commented:
@srikanthccv I pushed the fix for pylint issue.

I really hope it works now. I just cannot get it to run locally.

srikanthccv commented on CHANGELOG.md:
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- `opentelemetry-instrumentation-redis` Add `sanitize_query` config option to allow query sanitization. Enabled by default.
+  ([#1572](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1572))
```
```suggestion
- `opentelemetry-instrumentation-redis` Add `sanitize_query` config option to allow query sanitization.  ([#1572](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1572))
```

I believe this is not true.
